certbot-apache/certbot_apache/interfaces.py

```diff
@@ -186,6 +212,26 @@ class DirectiveNode(ParserNode):
     parameters: Tuple[str, ...]
 
     """
+    # pylint: disable=too-many-arguments, super-init-not-called
+    def __init__(self, name, parameters=(), ancestor=None, filepath="",
```

Comment:
I'm probably getting ahead of myself so you can ignore this comment for now if you prefer, but I was thinking we'll have to change these parameters a bit in the future because we need a way to pass in things like whitespace. With this in mind, maybe we don't need/want `name` and `parameter` options and it'll instead be passed through metadata. If we accept both, it's redundant and leaving ourselves open to the parameters differing.

This applies to the `BlockNode` as well.

Reply:
I think we should definitely address this, but in a later PR. However I've been thinking of the metadata as _everything else than what's already defined in the interface_ , and this is largely based on the idea I was fixed on, but now realize isn't relevant: storing the exact tokenized results of parsing within the ParserNode object. This would have meant that we change the AST bits only if the object is `dirty` upon `save()`. I see that this doesn't (most likely) stand true with the current plans we have.

However to keep everything simple and easy to access. I think what I'd personally prefer would be an additional method to `BlockNode` (the root node needs to be created "manually" anyway), something along the lines of `add_raw_child_block` that would take in the metadata instead of values for `name` and `parameters`. This would keep the interface easy and idiomatic to use, while providing a way for the internal parser functionality to initialize the objects in way that would preserve all the required information. I think this would be a rather small addition, but I'd prefer to do it in a separate PR in order to avoid growing this one larger.

Reply:
The alternative approach here would be initializing the instances with sequences including whitespaces - an approach I would personally prefer over having all this in metadata. 

My opinion mostly boils down to name and parameters being first class citizens that form the core of the whole node, while other information would and could belong to the metadata category.

Reply:
The alternative approach here would be initializing the instances with name and parameter sequences including whitespaces - an approach I would personally prefer over having all this in metadata. 

My opinion mostly boils down to name and parameters being first class citizens that form the core of the whole node, while other information would and could belong to the metadata category.